iroh/src/magicsock/endpoint_map/endpoint_state.rs

```diff
@@ -1137,39 +1145,160 @@ impl ConnectionState {
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns the number of paths.
+    pub fn len(&self) -> usize {
+        self.0.len()
     }
 }
 
 /// Information about a network path used by a [`Connection`].
 ///
+/// To get further information, like the latency, use [`Connection::path_stats`].
+///
 /// [`Connection`]: crate::endpoint::Connection
-#[derive(Debug, Clone, PartialEq, Eq)]
+/// [`Connection::path_stats`]: crate::endpoint::Connection::path_stats
+#[derive(derive_more::Debug, Clone)]
 pub struct PathInfo {
+    path_id: PathId,
+    #[debug(skip)]
+    handle: WeakConnectionHandle,
+    stats: PathStats,
+    remote: TransportAddr,
+    is_selected: bool,
+}
+
+impl PartialEq for PathInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.path_id == other.path_id
+            && self.remote == other.remote
+            && self.is_selected == other.is_selected
+    }
+}
+
+impl Eq for PathInfo {}
+
+impl PathInfo {
+    fn new(
+        path_id: PathId,
+        conn: &quinn::Connection,
+        remote: TransportAddr,
+        selected_path: Option<&TransportAddr>,
+    ) -> Option<Self> {
+        let stats = conn.path_stats(path_id)?;
+        Some(Self {
+            path_id,
+            handle: conn.weak_handle(),
+            is_selected: Some(&remote) == selected_path,
+            remote,
+            stats,
+        })
+    }
+
     /// The remote transport address used by this network path.
-    pub remote: TransportAddr,
-    /// The internal path identifier for the [`Connection`]
-    ///
-    /// This is unique for the lifetime of the connection.  Can be used to look up the path
-    /// statistics in the [`ConnectionStats::paths`], returned by [`Connection::stats`].
+    pub fn remote_addr(&self) -> &TransportAddr {
+        &self.remote
+    }
+
+    /// Returns `true` if this path is currently the main transmission path for this [`Connection`].
     ///
     /// [`Connection`]: crate::endpoint::Connection
-    /// [`Connection::stats`]: crate::endpoint::Connection::stats
-    /// [`ConnectionStats::paths`]: crate::endpoint::ConnectionStats::paths
-    // TODO: Decide if exposing this is a good idea.  Maybe we should just hide this
-    //    entirely try to provide Self::stats().  But that would mean this needs to have a
-    //    WeakConnectionHandle.
-    pub path_id: PathId,
+    pub fn is_selected(&self) -> bool {
+        self.is_selected
+    }
+
+    /// Whether this is an IP transport address.
+    pub fn is_ip(&self) -> bool {
+        self.remote.is_ip()
+    }
+
+    /// Whether this is a transport address via a relay server.
+    pub fn is_relay(&self) -> bool {
+        self.remote.is_relay()
+    }
+
+    /// Returns stats for this transmission path.
+    pub fn stats(&self) -> PathStats {
+        self.handle
+            .upgrade()
+            .and_then(|conn| conn.path_stats(self.path_id))
+            .unwrap_or(self.stats)
```

Comment:
Erm, you now just have these as a field?

Reply:
But they might be outdated. The field would only get refreshed when the open paths change.